[test] add tests to verify support for sending to bech32m address

sandipndev opened the pull request:
### Description

This PR is in reference to #396 
It adds some tests to confirm sending transactions to Bech32m addresses, introduced in [BIP350](https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki)

**Note:** This change requires `bitcoin-core` `v22.0` to be incorporated into bitcoind.

### Checklists

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

#### New Features:

* [x] I've added tests for Bech32m adoption by `rust-bitcoin`

RCasatta commented:
Ideally, I think it's a test to be added (also) to the `bdk_blockchain_tests!` macro

notmandatory commented:
Looks like bech32m addresses are not supported yet in `rust-bitcoin` 0.26.2. The PR that added support is https://github.com/rust-bitcoin/rust-bitcoin/pull/601 and it has other breaking changes that will also need to be fixed in `bdk`.  You can see the errors by updating `bdk`'s Cargo.toml like this:
```toml
bitcoin = { git = "https://github.com/rust-bitcoin/rust-bitcoin.git", branch = "master", features = ["use-serde", "base64"] }
```

Also @RCasatta has a good idea about also updating the `bdk_blockchain_tests` macro, so this could could be a bit larger task but also a good opportunity to learn about macros.  You'll also be able to get to know other parts of bdk by fixing the other errors to make `bdk` work with the `master` branch of `rust-bitcoin` (until a new released version is ready). Want to give it a go?

sandipndev commented:
Thank you for going through the problem.

We can develop using the master branch but should we release the Bech32m adoption before `rust-bitcoin` officially releases a version upgrade with the same?

In most places, there is a similar error after changing `Cargo.toml` to use the master branch of `rust-bitcoin`.
```bash
the trait `InnerXKey` is not implemented for `bitcoin::util::bip32`
```

I'm on fixing it!

notmandatory commented:
> We can develop using the master branch but should we release the Bech32m adoption before `rust-bitcoin` officially releases a version upgrade with the same?

You will need to keep this PR in "draft" mode until an official `rust-bitcoin` release is ready, and then you can put that version in the `Cargo.toml`. In the mean time you can still work on getting ready for that release by making your fixes against the `master` branch.

notmandatory commented:
The [0.27.0 version of `rust-bitcoin`](https://github.com/rust-bitcoin/rust-bitcoin/blob/master/CHANGELOG.md#027---2020-07-21) is now released which includes `bech32m` support. I was able to get your new test to pass after these steps:

  1. In `Cargo.toml` update `bitcoin` version to `"^0.27"`
  2. Fork and make local clones of the `rust-miniscript` and `rust-electrum-client` projects and update both also to  the `rust-bitcoin` version `0.27`
  3. In `Cargo.toml` update the dependencies there for `rust-miniscript` and `rust-electrum-client` to the local path of your updated clones of these project, something like this:
     ```
     miniscript = { path = "../rust-miniscript" }
     ...
     electrum-client = { path = "../rust-electrum-client", optional = true }
     ```
     
Give it a try and then we can discuss additional tests (as @RCasatta suggested) and how to make the updates to those other projects (if no one else is already working on it).

sandipndev commented:
For the time being, I changed the dependencies to point to my git fork with updated `rust-bitcoin v0.27` in order to check if all the test cases pass. 

The added test case for `bech32m` is passing with `cargo test`.

But it looks like builds and some other test cases are still failing.

RCasatta commented:
From a quick look, it seems you need to use patch in Cargo.toml for the rust-bitcoin 0.27 dep so that dependencies of dependencies are replaced as well, see https://doc.rust-lang.org/edition-guide/rust-2018/cargo-and-crates-io/replacing-dependencies-with-patch.html. 

Otherwise, you may end up for something that seems the same Object but they are from different crate versions

sandipndev committed:
Added a wallet unit test to send to a Bech32m address

sandipndev committed:
Added wallet blockchain test to send to Bech32m address

sandipndev commented:
Waiting for the following PRs to get merged:
- https://github.com/RCasatta/rust-bitcoincore-rpc/pull/1
- https://github.com/RCasatta/bitcoind/pull/29

notmandatory commented:
I created a couple more PRs that also need to be merged to support this PR:

1. https://github.com/RCasatta/bitcoind/pull/30
2. https://github.com/RCasatta/electrsd/pull/21

~~1. https://github.com/RCasatta/rust-bitcoincore-rpc/pull/2~~ (optional, not really needed for this PR)

sandipndev commented on src/testutils/blockchain_tests.rs:
```diff
@@ -875,6 +873,53 @@ macro_rules! bdk_blockchain_tests {
+                            ImportDescriptorRequest::new(wallet_descriptor, false),
+                            ImportDescriptorRequest::new(change_descriptor, false),
+                        ]
+                    ).unwrap();
```
Updated this section to match the newer prototype of `import_descriptors` in `rust-bitcoincore-rpc`

notmandatory commented:
Track via #63

notmandatory commented:
This needs to be reworked a little to use raw rpc calls to create our test `tr()` descriptor wallet as @RCasatta suggested. Once https://github.com/rust-bitcoin/rust-bitcoincore-rpc/pull/174 is merged we can rework the tests to use the lib instead.

Moving to '0.15.0' feature freeze so we have more time to work on it.

notmandatory commented:
I got the blockchain_tests to work without the `core-rpc` or `bitcoind` patches :tada:. Once I have my code cleaned up I'll add it as a commit to this PR. Moving this back into code freeze for `0.14.0`.

notmandatory committed:
Fix blockchain_tests::test_send_to_bech32m_addr

Now works with latest released versions of rust-bitcoincore-rpc and
bitcoind. Once these crates are updated to support creating descriptor
wallets and add importdescriptors and bech32m support this test will
need to be updated.

notmandatory reviewed (approved):
ACK afa1ab4ff8dba441d30bc745eca2d634fbd7f017

notmandatory commented:
FYI, here is an explanation from stackexchange that we followed for creating a P2TR wallet in `bitcoind` and then creating a deposit address for it: https://bitcoin.stackexchange.com/questions/108006/how-to-make-a-taproot-transaction-with-bitcoin-cli

sandipndev commented:
Awesome, thanks for fixing this PR. Calling the RPC commands, for now, seems to work perfectly and tests run - which is fantastic and ACK https://github.com/bitcoindevkit/bdk/pull/402/commits/afa1ab4ff8dba441d30bc745eca2d634fbd7f017 🎉

rajarshimaitra commented on src/testutils/blockchain_tests.rs:
```diff
@@ -145,9 +145,7 @@ impl TestClient {
-
-        let monitor_script =
-            tx.vout[0].script_pub_key.addresses.as_ref().unwrap()[0].script_pubkey();
+        let monitor_script = Script::from_hex(&mut tx.vout[0].script_pub_key.hex.to_hex()).unwrap();
```
Is there any special benefit of getting the script this way?

notmandatory replied:
I believe this was required because `.script_pubkey()` function didn't support P2TR scripts.

rajarshimaitra commented on src/testutils/blockchain_tests.rs:
```diff
@@ -911,6 +909,102 @@ macro_rules! bdk_blockchain_tests {
+
+                // TODO replace once rust-bitcoincore-rpc with PR 174 released
+                // https://github.com/rust-bitcoin/rust-bitcoincore-rpc/pull/174
+                let _createwallet_result: Value = test_client.bitcoind.client.call("createwallet", &["taproot_wallet".into(),false.into(),true.into(),serde_json::to_value("").unwrap(), false.into(), true.into()]).unwrap();
```
It seems we don't need to hold the create wallet result. So we can ignore the let binding here.

sandipndev replied:
No, I think this could be removed out

notmandatory replied:
I'm going to leave the unused results in for now since they don't hurt anything and we can remove them when we redo this test when updated dependencies are ready.

rajarshimaitra commented on src/testutils/blockchain_tests.rs:
```diff
@@ -911,6 +909,102 @@ macro_rules! bdk_blockchain_tests {
+                        ];
+
+                // TODO replace once rust-bitcoincore-rpc with PR 199 released
+                let _import_result: Value = taproot_wallet_client.call("importdescriptors", &[serde_json::to_value(tr_descriptors).unwrap()]).unwrap();
```
Same as before, no need to hold the result.

rajarshimaitra reviewed (commented):
tACK https://github.com/bitcoindevkit/bdk/pull/402/commits/afa1ab4ff8dba441d30bc745eca2d634fbd7f017.

below are few minor question and suggestions.
